File: lazerapp/public/js/purchase_receipt.js
// frappe.ui.form.on("Purchase Receipt", {
//     refresh(frm) {
//         if (frm.doc.docstatus === 1) {
//             frm.add_custom_button("Make GRN Payment", () => {
//                 frappe.new_doc("GRN Payment", {
//                     supplier: frm.doc.supplier,
//                     company: frm.doc.company,
//                     purchase_receipt: frm.doc.name,
//                     amount: frm.doc.grand_total
//                 });
//             }).addClass("btn-success");
//         }
//     }
// });



/********************************
 * FINAL ITEM CALCULATION (CORRECT)
 ********************************/
frappe.ui.form.on('Purchase Receipt Item', {
    custom_quantity(frm, cdt, cdn) {
        calculate_item(frm, cdt, cdn);
    },

    rate(frm, cdt, cdn) {
        calculate_item(frm, cdt, cdn);
    },

    custom_vat_bd(frm, cdt, cdn) {
        calculate_item(frm, cdt, cdn);
    }
});

function calculate_item(frm, cdt, cdn) {
    let row = locals[cdt][cdn];

    let qty = flt(row.custom_quantity);
    let rate = flt(row.rate);
    let vat = flt(row.custom_vat_bd);

    let amount = (qty * rate) + vat;

    frappe.model.set_value(cdt, cdn, "amount", amount);

    frm.refresh_field("items");

    // Let ERPNext recalculate all totals
    frm.trigger("calculate_taxes_and_totals");
}

File: lazerapp/doctype_overrides/purchase_receipt.py
import frappe
from erpnext.stock.doctype.purchase_receipt.purchase_receipt import PurchaseReceipt


class CustomPurchaseReceipt(PurchaseReceipt):

    def before_save(self):
        self.update_amount_with_vat()

    def before_submit(self):
        """
        FINAL sync before stock posting
        """
        self.sync_qty_with_custom_quantity()
        self.update_amount_with_vat()

    def sync_qty_with_custom_quantity(self):
        """
        Ensure ERPNext uses correct quantity on submit
        """
        for item in self.items:
            if item.custom_quantity:
                item.qty = item.custom_quantity

    def update_amount_with_vat(self):
        """
        amount = (rate * custom_quantity) + custom_vat_bd
        """
        total = 0.0

        for item in self.items:
            rate = float(item.rate or 0)
            qty = float(item.custom_quantity or 0)
            vat = float(item.custom_vat_bd or 0)

            item.amount = (rate * qty) + vat
            total += item.amount

        self.total = total
        self.grand_total = total
        self.base_grand_total = total
